scripts/check_docstring_coverage.py

```diff
@@ -0,0 +1,158 @@
+    for file, func_list in all_funcs.items():
+        public_funcs: List[ast.FunctionDef] = list(
+            filter(
+                lambda f: _function_filter(f),
+                func_list,
+            )
+        )
+        for func in public_funcs:
+            result: Tuple[ast.FunctionDef, bool] = (func, bool(ast.get_docstring(func)))
+            diagnostics[file].append(result)
+
+    return diagnostics
+
+
+def _function_filter(func: ast.FunctionDef) -> bool:
+    # Private and dunder funcs/methods
+    if func.name.startswith("_"):
+        return False
+    # Getters and setters
+    for decorator in func.decorator_list:
+        if (isinstance(decorator, ast.Name) and decorator.id == "property") or (
+            isinstance(decorator, ast.Attribute) and decorator.attr == "setter"
+        ):
+            return False
+    return True
+
+
+def review_diagnostics(diagnostics: Diagnostics, changed_files: List[str]) -> None:
+    """Generate the report to stdout.
+
+    Args:
+        diagnostics (Diagnostics): The diagnostics generated in `gather_docstring_diagnostics`.
+        changed_files (List[str]): The list of files generated from `get_changed_files`.
+
+    Raises:
+        AssertionError if threshold is surpassed. This threshold ensures we don't introduce new regressions.
+    """
+    total_passed: int = 0
+    total_funcs: int = 0
+
+    relevant_diagnostics: Dict[str, List[ast.FunctionDef]] = defaultdict(list)
+
+    for file, diagnostics_list in diagnostics.items():
+        relevant_file: bool = file in changed_files
+        for func, success in diagnostics_list:
+            if success:
+                total_passed += 1
+            elif not success and relevant_file:
+                relevant_diagnostics[file].append(func)
+            total_funcs += 1
+
+    print(
+        f"[SUMMARY] {total_passed} of {total_funcs} public functions ({100 * total_passed / total_funcs:.2f}%) have docstrings!"
+    )
+
+    if relevant_diagnostics:
+        print(
+            "\nHere are violations of the style guide that are relevant to the files changed in your PR:"
+        )
+
+        for file, func_list in relevant_diagnostics.items():
+            print(f"  {file}:")
+            for func in func_list:
+                print(f"    L{func.lineno}:{func.name}")
+
+    # Chetan - 20220305 - While this number should be 0, getting the number of style guide violations down takes time
+    # and effort. In the meanwhile, we want to set an upper bound on errors to ensure we're not introducing
+    # further regressions. As docstrings are added, developers should update this number.
+    total_failed: int = total_funcs - total_passed
+    assert (
+        total_failed <= DOCSTRING_ERROR_THRESHOLD
+    ), f"""A public function without a docstring was introduced; please resolve the matter before merging.
+                We expect there to be {total_failed} or fewer violations of the style guide (actual: {total_failed})"""
```

Comment:
total_failed should not appear twice (only in the actual: ...) area. Also, right now the threshold is below the actual, if merged this check would always fail

Reply:
good call! should be up to date now